File: PHP/2818. Apply Operations to Maximize Score.php
<?php
/*
You are given an array nums of n positive integers and an integer k.
Initially, you start with a score of 1. You have to maximize your score by applying the following operation at most k times:
Choose any non-empty subarray nums[l, ..., r] that you haven't chosen previously.
Choose an element x of nums[l, ..., r] with the highest prime score. If multiple such elements exist, choose the one with the smallest index.
Multiply your score by x.
Here, nums[l, ..., r] denotes the subarray of nums starting at index l and ending at the index r, both ends being inclusive.
The prime score of an integer x is equal to the number of distinct prime factors of x. 
    For example, the prime score of 300 is 3 since 300 = 2 * 2 * 3 * 5 * 5.
Return the maximum possible score after applying at most k operations.
Since the answer may be large, return it modulo 109 + 7.
*/
class Solution {
    const MOD = 1000000007;
    /**
     * 2818. Apply Operations to Maximize Score
     * @param Integer[] $nums
     * @param Integer $k
     * @return Integer
     */
    function maximumScore($nums, $k) {
        $n = count($nums);

        // Step 1: Prime scores
        $primeScores = array_map([$this, 'countDistinctPrimes'], $nums);

        // Step 2: Monotonic stack to find left and right boundaries
        $left = array_fill(0, $n, -1);
        $right = array_fill(0, $n, $n);
        $stack = [];

        for ($i = 0; $i < $n; $i++) {
            while (!empty($stack) && $primeScores[$stack[count($stack) - 1]] < $primeScores[$i]) {
                $idx = array_pop($stack);
                $right[$idx] = $i;
            }
            $left[$i] = empty($stack) ? -1 : $stack[count($stack) - 1];
            $stack[] = $i;
        }

        // Step 3: Sort indices by nums descending
        $indices = range(0, $n - 1);
        usort($indices, function($a, $b) use ($nums) {
            return $nums[$b] <=> $nums[$a];
        });

        // Step 4: Apply operations greedily
        $score = 1;
        foreach ($indices as $i) {
            $count = ($i - $left[$i]) * ($right[$i] - $i);
            $take = min($k, $count);
            $score = ($score * $this->modPow($nums[$i], $take, self::MOD)) % self::MOD;
            $k -= $take;
            if ($k == 0) break;
        }

        return $score;
    }

    private function modPow($base, $exp, $mod) {
        $result = 1;
        $base %= $mod;
        while ($exp > 0) {
            if ($exp % 2 === 1) {
                $result = ($result * $base) % $mod;
            }
            $base = ($base * $base) % $mod;
            $exp = intdiv($exp, 2);
        }
        return $result;
    }

    private function countDistinctPrimes($num) {
        $count = 0;
        if ($num % 2 === 0) {
            $count++;
            while ($num % 2 === 0) $num = intdiv($num, 2);
        }
        for ($i = 3; $i * $i <= $num; $i += 2) {
            if ($num % $i === 0) {
                $count++;
                while ($num % $i === 0) $num = intdiv($num, $i);
            }
        }
        if ($num > 1) $count++;
        return $count;
    }
}

$c = new Solution;
print $c->maximumScore([8,3,9,3,8], 2)."\n"; // Expect 81
print $c->maximumScore([19,12,14,6,10,18], 3)."\n"; // Expect 4788

?>

File: Swift/2818. Apply Operations to Maximize Score.swift
/*
You are given an array nums of n positive integers and an integer k.
Initially, you start with a score of 1. You have to maximize your score by applying the following operation at most k times:
Choose any non-empty subarray nums[l, ..., r] that you haven't chosen previously.
Choose an element x of nums[l, ..., r] with the highest prime score. If multiple such elements exist, choose the one with the smallest index.
Multiply your score by x.
Here, nums[l, ..., r] denotes the subarray of nums starting at index l and ending at the index r, both ends being inclusive.
The prime score of an integer x is equal to the number of distinct prime factors of x. 
    For example, the prime score of 300 is 3 since 300 = 2 * 2 * 3 * 5 * 5.
Return the maximum possible score after applying at most k operations.
Since the answer may be large, return it modulo 109 + 7.
    
2818. Apply Operations to Maximize Score    
*/
class Solution {
    let MOD = Int(1e9 + 7)
    func maximumScore(_ nums: [Int], _ k: Int) -> Int {
        let n = nums.count
        let primeScores = nums.map { countDistinctPrimes($0) }

        // Step 1: Monotonic stack to find left and right boundaries
        var left = Array(repeating: -1, count: n)
        var right = Array(repeating: n, count: n)
        var stack: [Int] = []

        for i in 0..<n {
            while let last = stack.last, primeScores[last] < primeScores[i] {
                right[stack.removeLast()] = i
            }
            left[i] = stack.last ?? -1
            stack.append(i)
        }

        // Step 2: Sort indices by nums[i] in descending order
        let indices = (0..<n).sorted { nums[$0] > nums[$1] }

        // Step 3: Apply operations greedily
        var score = 1
        var remainingK = k

        for i in indices {
            let count = (i - left[i]) * (right[i] - i)
            let take = min(remainingK, count)
            score = modPow(score, 1, MOD)
            score = (score * modPow(nums[i], take, MOD)) % MOD
            remainingK -= take
            if remainingK == 0 { break }
        }

        return score
    }

    private func modPow(_ base: Int, _ exp: Int, _ mod: Int) -> Int {
        var result = 1
        var base = base % mod
        var exp = exp

        while exp > 0 {
            if exp % 2 == 1 {
                result = (result * base) % mod
            }
            base = (base * base) % mod
            exp /= 2
        }
        return result
    }

    private func countDistinctPrimes(_ num: Int) -> Int {
        var x = num
        var count = 0
        if x % 2 == 0 {
            count += 1
            while x % 2 == 0 {
                x /= 2
            }
        }
        var i = 3
        while i * i <= x {
            if x % i == 0 {
                count += 1
                while x % i == 0 {
                    x /= i
                }
            }
            i += 2
        }
        if x > 1 {
            count += 1
        }
        return count
    }
}
let solution = Solution()
print(solution.maximumScore([8,3,9,3,8], 2)) // Expect 81
print(solution.maximumScore([19,12,14,6,10,18], 3))  // Expect 4788